src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/JTBestHaplotype.java

```diff
@@ -0,0 +1,260 @@
+        // Standard behavior from the old GraphBasedKBestHaplotypeFinder, base our next path on the edge weights instead
+        int totalOutgoingMultiplicity = 0;
+        for (final BaseEdge edge : outgoingEdges) {
+            totalOutgoingMultiplicity += edge.getMultiplicity();
+        }
+
+        // Add all valid edges to the graph
+        for (final E edge : outgoingEdges) {
+            // Don't traverse an edge if it only has reference evidence supporting it (unless there is no other evidence whatsoever)
+            if (!edgesAccountedForByJunctionTrees.contains((MultiSampleEdge) edge) &&
+                    totalOutgoingMultiplicity != 0 && edge.getMultiplicity() != 0) {
+
+                // TODO better justify this to myself and others
+                // only include ref edges with multiplicity of 1 (i.e. only the ref read spanned it) if there are no other choices at this site (from Junction trees or otherwise)
+                if ((edge.isRef() && edge.getMultiplicity() == 1) &&
+                        !(edgesAccountedForByJunctionTrees.isEmpty() && outgoingEdges.size() < 2)) { // no junction tree evidence and only one ref path to take
+                    continue;
+                }
+
+                List<E> chainCopy = new ArrayList<>(chain);
+                chainCopy.add(edge);
+                output.add(new JTBestHaplotype<>(this, chainCopy, edge.getMultiplicity(), totalOutgoingMultiplicity, false));
+            }
+        }
+        return output;
+    }
+
+    /**
+     * Return an accounting of how many edge decision (where there is a choice) have consecutively been based on the raw graph weights
+     * (as opposed to being based on junction tree evidence).
+     *
+     * @return number of decision edges
+     */
+    public int getDecisionEdgesTakenSinceLastJunctionTreeEvidence() {
+        return decisionEdgesTakenSinceLastJunctionTreeEvidence;
+    }
+
+    /**
+     * Add a junction tree (corresponding to the current vertex for traversal, note that if a tree has already been visited by this path then it is ignored)
+     * @param junctionTreeForNode Junction tree to add
+     */
+    public void addJunctionTree(final ExperimentalReadThreadingGraph.ThreadingTree junctionTreeForNode) {
+        if (treesInQueue.addJunctionTree(junctionTreeForNode)) {
+            decisionEdgesTakenSinceLastJunctionTreeEvidence = 0;
+        }
+    }
+
+    /**
+     * A helper class for managing the various junction tree operations that need to be done by JTBestHaplotypeFinder
+     */
+    private class JunctionTreeView {
+        Set<ExperimentalReadThreadingGraph.ThreadingTree> visitedTrees;
+        List<ExperimentalReadThreadingGraph.ThreadingNode> activeNodes;
+
+        protected JunctionTreeView() {
+            visitedTrees = new HashSet<>();
+            activeNodes = new ArrayList<>(5);
```

Comment:
Hmm... The problem with Queueus is that they don't support arbitrary access and we need to be able to access trees besides the fist one. 

Also LinkedList queues are often not very much faster because of the dereferecing/traversal cost for access.